README.md
# Urbica React Mapbox GL Cluster

[![Build Status](https://img.shields.io/circleci/project/github/urbica/react-map-gl-cluster.svg?style=popout)](https://circleci.com/gh/urbica/react-map-gl-cluster) [![Greenkeeper badge](https://badges.greenkeeper.io/urbica/react-map-gl-cluster.svg)](https://greenkeeper.io/)
![npm](https://img.shields.io/npm/dt/@urbica/react-map-gl-cluster.svg)
![npm](https://img.shields.io/npm/v/@urbica/react-map-gl-cluster.svg)

Cluster component for [Urbica React Components Library for Mapbox GL JS](https://github.com/urbica/react-map-gl).

## Installation

```shell
npm install mapbox-gl supercluster @urbica/react-map-gl @urbica/react-map-gl-cluster
```

...or if you are using yarn:

```shell
yarn add mapbox-gl supercluster @urbica/react-map-gl @urbica/react-map-gl-cluster
```

## Usage

```jsx
import { randomPoint } from '@turf/random';
import MapGL, { Marker } from '@urbica/react-map-gl';
import Cluster from '@urbica/react-map-gl-cluster';
import 'mapbox-gl/dist/mapbox-gl.css';

const bbox = [-160, -70, 160, 70];
const points = randomPoint(50, { bbox }).features;
points.forEach((point, index) => (point.id = index));

initialState = {
  viewport: {
    latitude: 0,
    longitude: 0,
    zoom: 0
  }
};

const style = {
  width: '20px',
  height: '20px',
  color: '#fff',
  background: '#1978c8',
  borderRadius: '20px',
  textAlign: 'center'
};

const ClusterMarker = ({ longitude, latitude, pointCount }) => (
  <Marker longitude={longitude} latitude={latitude}>
    <div style={{ ...style, background: '#f28a25' }}>{pointCount}</div>
  </Marker>
);

<MapGL
  style={{ width: '100%', height: '400px' }}
  mapStyle='mapbox://styles/mapbox/light-v9'
  accessToken={MAPBOX_ACCESS_TOKEN}
  onViewportChange={viewport => setState({ viewport })}
  {...state.viewport}
>
  <Cluster radius={40} extent={512} nodeSize={64} component={ClusterMarker}>
    {points.map(point => (
      <Marker
        key={point.id}
        longitude={point.geometry.coordinates[0]}
        latitude={point.geometry.coordinates[1]}
      >
        <div style={style} />
      </Marker>
    ))}
  </Cluster>
</MapGL>;
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [0.2.0](https://github.com/urbica/react-map-gl-cluster/compare/v0.1.0...v0.2.0) (2019-10-24)


### Features

* add map and reduce props according supercluster api ([916a4f0](https://github.com/urbica/react-map-gl-cluster/commit/916a4f08a41148825c53ad507a3a31d45a2cbc39))
* add map and reduce props according supercluster api [#25](https://github.com/urbica/react-map-gl-cluster/issues/25) (h/t (Akiyamka)[https://github.com/Akiyamka]) ([a6e37d7](https://github.com/urbica/react-map-gl-cluster/commit/a6e37d71e7ad6d4678889a06c9c4980facafaba6))


### Bug Fixes

* ClusterReduceFunction type ([b04cefb](https://github.com/urbica/react-map-gl-cluster/commit/b04cefb85e51011fce28aa31bbe91805b7a87cdf))

# 0.1.0 (2019-03-11)


### Features

* initial commit ([6a3414b](https://github.com/urbica/react-map-gl-cluster/commit/6a3414b))


src/components/Cluster/index.test.js
import React from 'react';
import { mount } from 'enzyme';
import MapGL, { Marker } from '@urbica/react-map-gl';
import Cluster from '../..';

test('Cluster#render', () => {
  const Element = () => <div>ok</div>;
  const ClusterComponent = () => <div>ok</div>;

  const wrapper = mount(
    <MapGL latitude={0} longitude={0} zoom={0}>
      <Cluster component={ClusterComponent}>
        <Marker longitude={0} latitude={0}>
          <Element />
        </Marker>
        <Marker longitude={1} latitude={1}>
          <Element />
        </Marker>
        <Marker longitude={2} latitude={2}>
          <Element />
        </Marker>
      </Cluster>
    </MapGL>
  );

  expect(wrapper.find('Cluster').exists()).toBe(true);
  expect(wrapper.find('ClusterComponent').exists()).toBe(true);
  expect(wrapper.find('ClusterComponent')).toHaveLength(1);

  wrapper.unmount();
  expect(wrapper.find('Cluster').exists()).toBe(false);
  expect(wrapper.find('ClusterComponent').exists()).toBe(false);
});


styleguide.config.js
const path = require('path');
const webpack = require('webpack');

module.exports = {
  title: 'Urbica React Map GL Cluster',
  usageMode: 'expand',
  exampleMode: 'expand',
  pagePerSection: true,
  require: [path.resolve(__dirname, 'styleguide.setup.js')],
  moduleAliases: {
    '@urbica/react-map-gl-cluster': path.resolve(__dirname, 'src')
  },
  webpackConfig: {
    module: {
      rules: [
        {
          test: /\.jsx?$/,
          exclude: /node_modules/,
          loader: 'babel-loader'
        },
        {
          test: /\.css$/,
          use: ['style-loader', 'css-loader']
        }
      ]
    }
  },
  dangerouslyUpdateWebpackConfig: (webpackConfig) => {
    const envPlugin = new webpack.EnvironmentPlugin(['MAPBOX_ACCESS_TOKEN']);
    webpackConfig.plugins.push(envPlugin);
    return webpackConfig;
  }
};


src/components/Cluster/index.js
// @flow

import Supercluster from 'supercluster';
import { Children, PureComponent, createElement } from 'react';
import { MapContext } from '@urbica/react-map-gl';
import type MapboxMap from 'mapbox-gl/src/ui/map';

import point from '../../utils/point';
import shallowCompareChildren from '../../utils/shallowCompareChildren';

export type SuperclusterFeature = {
  type: 'Feature',
  id: number,
  properties: {
    cluster: true,
    cluster_id: number,
    point_count: number,
    point_count_abbreviated: string | number
  },
  geometry: {
    type: 'Point',
    coordinates: [number, number]
  }
};

export type ClusterComponentProps = {
  longitude: number,
  latitude: number,
  clusterId: number,
  pointCount: number,
  pointCountAbbreviated: string | number
};

export type ClusterMapFunction = (props: {}) => any;

export type ClusterReduceFunction = (accumulated: {}, props: {}) => void;

export type ClusterComponent = React$Component<ClusterComponentProps, any>;

type Props = {
  /** Minimum zoom level at which clusters are generated */
  minZoom?: number,

  /** Maximum zoom level at which clusters are generated */
  maxZoom?: number,

  /** Cluster radius, in pixels */
  radius?: number,

  /** (Tiles) Tile extent. Radius is calculated relative to this value */
  extent?: number,

  /** Size of the KD-tree leaf node. Affects performance */
  nodeSize?: number,

  /**
   * A function that returns cluster properties
   * corresponding to a single point.
   *  */
  // eslint-disable-next-line react/no-unused-prop-types
  map?: ClusterMapFunction,

  /** A reduce function that merges properties of two clusters into one. */
  // eslint-disable-next-line react/no-unused-prop-types
  reduce?: ClusterReduceFunction,

  /** React Component for rendering Cluster */
  component: Class<ClusterComponent>,

  /** List of Markers */
  children: React$Node
};

type State = {
  clusters: Array<Object>
};

class Cluster extends PureComponent<Props, State> {
  _map: MapboxMap;

  _cluster: Object;

  static displayName = 'Cluster';

  static defaultProps = {
    minZoom: 0,
    maxZoom: 16,
    radius: 40,
    extent: 512,
    nodeSize: 64
  };

  constructor(props: Props) {
    super(props);

    this.state = {
      clusters: []
    };
  }

  componentDidMount() {
    this._createCluster(this.props);
    this._recalculate();

    this._map.on('moveend', this._recalculate);
  }

  componentDidUpdate(prevProps: Props) {
    const shouldUpdate =
      prevProps.minZoom !== this.props.minZoom ||
      prevProps.maxZoom !== this.props.maxZoom ||
      prevProps.radius !== this.props.radius ||
      prevProps.extent !== this.props.extent ||
      prevProps.nodeSize !== this.props.nodeSize ||
      !shallowCompareChildren(prevProps.children, this.props.children);

    if (shouldUpdate) {
      this._createCluster(this.props);
      this._recalculate();
    }
  }

  componentWillUnmount() {
    if (!this._map || !this._map.getStyle()) {
      return;
    }

    this._map.off('moveend', this._recalculate);
  }

  getCluster() {
    return this._cluster;
  }

  _createCluster = (props: Props) => {
    const {
      minZoom,
      maxZoom,
      radius,
      extent,
      nodeSize,
      children,
      reduce,
      map
    } = props;

    const cluster = new Supercluster({
      minZoom,
      maxZoom,
      radius,
      extent,
      nodeSize,
      reduce,
      map
    });

    const points = Children.map(children, child =>
      point([child.props.longitude, child.props.latitude], child)
    );

    cluster.load(points);
    this._cluster = cluster;
  };

  _recalculate = () => {
    const zoom = this._map.getZoom();
    const bounds = this._map.getBounds().toArray();
    const bbox = bounds[0].concat(bounds[1]);

    const clusters = this._cluster.getClusters(bbox, Math.round(zoom));
    this.setState(() => ({ clusters }));
  };

  _renderCluster = (cluster: SuperclusterFeature) => {
    const [longitude, latitude] = cluster.geometry.coordinates;
    const {
      cluster_id: clusterId,
      point_count: pointCount,
      point_count_abbreviated: pointCountAbbreviated
    } = cluster.properties;

    return createElement(this.props.component, {
      longitude,
      latitude,
      clusterId,
      pointCount,
      pointCountAbbreviated,
      key: `cluster-${cluster.properties.cluster_id}`
    });
  };

  render() {
    return createElement(MapContext.Consumer, {}, (map) => {
      if (map) {
        this._map = map;
      }

      if (this.state.clusters.length === 0) {
        return null;
      }

      const clusters = this.state.clusters.map((cluster) => {
        if (cluster.properties.cluster) {
          return this._renderCluster(cluster);
        }
        const { type, key, props } = cluster.properties;
        return createElement(type, { key, ...props });
      });

      return clusters;
    });
  }
}

export default Cluster;
